src/core/basics.ml

```diff
@@ -217,29 +204,28 @@ let has_metas : bool -> term -> bool =
   let exception Found in fun b t ->
   try iter_meta b (fun _ -> raise Found) t; false with Found -> true
 
-(** [distinct_vars_opt ts] checks that [ts] is made of distinct
+(** [distinct_vars ts] checks that [ts] is made of distinct
    variables and returns these variables. *)
-let distinct_vars_opt : term array -> tvar array option =
-  let exception Not_unique_var in fun ts ->
+let distinct_vars : ctxt -> term array -> tvar array option =
+  let exception Not_unique_var in fun ctx ts ->
   let open Pervasives in
   let vars = ref VarSet.empty in
   let to_var t =
     match unfold t with
-    | Vari x when not (VarSet.mem x !vars) -> vars := VarSet.add x !vars; x
-    | _ -> raise Not_unique_var
+    | Vari(x) ->
+        let x =
+          (* If variable is defined in [ctx], replace the var by its
+             definition. *)
+          match Ctxt.def_of x ctx with
+          | Some(Vari(x)) -> x
+          | None          -> x
+          | Some(_)       -> raise Not_unique_var
+        in
+        if not (VarSet.mem x !vars) then (vars := VarSet.add x !vars; x) else
+        raise Not_unique_var
+    | _       -> raise Not_unique_var
   in try Some (Array.map to_var ts) with Not_unique_var -> None
```

Comment:
I think that here `to_var` needs to be recursive: what if `x` is defined to be `y`, and `y` is defined to be `z`?

Also, please put the final `try ... with ...` on its own line.

Reply:
Indeed.

Reply:
Shouldn't we include let's unfolding in the (recursive) unfold function instead? unfold could take an optional (default []) context as first argument.

Reply:
I was thinking about that, but it would create a dependency between unfold and contexts

Reply:
No, I don't think `unfold` should do that. There are many cases where unfolding is not the desired behaviour. And that would also make it less efficient, and `unfold` is really critical.